plugin/role_set.go

```diff
@@ -93,319 +106,259 @@ func (rs *RoleSet) bindingHash() string {
+		return nil, fmt.Errorf("couldn't fetch service account %q: {{err}}", accountId.ResourceName())
 	}
 
 	return account, nil
 }
 
-type ResourceBindings map[string]util.StringSet
+// saveRoleSetWithNewAccount rotates the role set service account. This includes creating a new service account with
+// a new name and deleting the old service account, updating keys or bindings as required.
+func (b *backend) saveRoleSetWithNewAccount(ctx context.Context, req *logical.Request, rs *RoleSet, project string, newBinds ResourceBindings, scopes []string) (warnings []string, err error) {
+	b.Logger().Debug("updating roleset with new account")
 
-func (rb ResourceBindings) asOutput() map[string][]string {
-	out := make(map[string][]string)
-	for k, v := range rb {
-		out[k] = v.ToSlice()
-	}
-	return out
-}
-
-type TokenGenerator struct {
-	KeyName    string
-	B64KeyJSON string
+	oldResources := rs.boundResources()
 
-	Scopes []string
-}
+	// Generate name for new account
+	newSaName := generateAccountNameForRoleSet(rs.Name)
 
-func (b *backend) saveRoleSetWithNewAccount(ctx context.Context, s logical.Storage, rs *RoleSet, project string, newBinds ResourceBindings, scopes []string) (warning []string, err error) {
-	b.rolesetLock.Lock()
-	defer b.rolesetLock.Unlock()
-
-	httpC, err := b.HTTPClient(s)
-	if err != nil {
-		return nil, err
+	// Construct IDs for new resources.
+	// The actual GCP resources are not created yet, but we need the IDs to create WAL entries.
+	newResources := &gcpAccountResources{
+		accountId: gcputil.ServiceAccountId{
+			Project:   project,
+			EmailOrId: emailForServiceAccountName(project, newSaName),
+		},
+		bindings: newBinds,
+	}
+	if len(scopes) > 0 {
+		newResources.tokenGen = &TokenGenerator{Scopes: scopes}
 	}
 
-	iamAdmin, err := b.IAMAdminClient(s)
+	// Add WALs for both old and new resources.
+	// WAL callback checks whether resources are still being used by roleset so
+	// there is no harm in adding WALs early, or adding WALs for resources that
+	// will eventually get cleaned up.
+	b.Logger().Debug("adding WALs for old roleset resources")
+	oldWalIds, err := b.addWalsForRoleSetResources(ctx, req, rs.Name, oldResources)
 	if err != nil {
 		return nil, err
 	}
 
-	apiHandle := iamutil.GetApiHandle(httpC, useragent.String())
-
-	oldAccount := rs.AccountId
-	oldBindings := rs.Bindings
-	oldTokenKey := rs.TokenGen
-
-	oldWals, err := rs.addWALsForCurrentAccount(ctx, s)
+	b.Logger().Debug("adding WALs for new roleset resources")
+	newWalIds, err := b.addWalsForRoleSetResources(ctx, req, rs.Name, newResources)
 	if err != nil {
-		tryDeleteWALs(ctx, s, oldWals...)
-		return nil, errwrap.Wrapf("failed to create WAL for cleaning up old account: {{err}}", err)
+		return nil, err
 	}
 
-	newWals := make([]string, 0, len(newBinds)+2)
-	walId, err := rs.newServiceAccount(ctx, s, iamAdmin, project)
+	// Created new RoleSet resources
```

Comment:
What is this referring to?

Reply:
This was copied verbatim from #62.

I am guessing this is referring to the bunch of GCP resources that will have to be created for this roleset, including:

- GCP Service Account
- IAM Bindings
- Service Account key if this was an `access_token`.